sdks/java/io/mongodb/src/main/java/org/apache/beam/sdk/io/mongodb/MongoDbGridFSIO.java

```diff
@@ -446,4 +504,164 @@ public void close() throws IOException {
+     */
+    void write(T output, OutputStream outStream) throws IOException;
+  }
+
+  /**
+   * A {@link PTransform} to write data to MongoDB GridFS.
+   */
+  @AutoValue
+  public abstract static class Write<T> extends PTransform<PCollection<T>, PDone> {
+    @Nullable abstract String uri();
+    @Nullable abstract String database();
+    @Nullable abstract String bucket();
+    @Nullable abstract Long chunkSize();
+    abstract WriteFn<T> writeFn();
+    @Nullable abstract String filename();
+
+    abstract Builder<T> toBuilder();
+
+    @AutoValue.Builder
+    abstract static class Builder<T> {
+      abstract Builder<T> setUri(String uri);
+      abstract Builder<T> setDatabase(String database);
+      abstract Builder<T> setBucket(String bucket);
+      abstract Builder<T> setFilename(String filename);
+      abstract Builder<T> setChunkSize(Long chunkSize);
+      abstract Builder<T> setWriteFn(WriteFn<T> fn);
+      abstract Write<T> build();
+    }
+
+    public Write<T> withUri(String uri) {
+      checkNotNull(uri);
+      return toBuilder().setUri(uri).build();
+    }
+
+    public Write<T> withDatabase(String database) {
+      checkNotNull(database);
+      return toBuilder().setDatabase(database).build();
+    }
+
+    public Write<T> withBucket(String bucket) {
+      checkNotNull(bucket);
+      return toBuilder().setBucket(bucket).build();
+    }
+
+    public Write<T> withFilename(String filename) {
+      checkNotNull(filename);
+      return toBuilder().setFilename(filename).build();
+    }
+
+    public Write<T> withChunkSize(Long chunkSize) {
+      checkNotNull(chunkSize);
+      checkArgument(chunkSize > 1, "Chunk Size must be greater than 1", chunkSize);
+      return toBuilder().setChunkSize(chunkSize).build();
+    }
+
+    public void validate(T input) {
+      checkNotNull(filename(), "filename");
+      checkNotNull(writeFn(), "writeFn");
+    }
+
+    @Override
+    public void populateDisplayData(DisplayData.Builder builder) {
+      super.populateDisplayData(builder);
+      builder.addIfNotNull(DisplayData.item("uri", uri()));
+      builder.addIfNotNull(DisplayData.item("database", database()));
+      builder.addIfNotNull(DisplayData.item("bucket", bucket()));
+      builder.addIfNotNull(DisplayData.item("filename", filename()));
+    }
+
+    @Override
+    public PDone apply(PCollection<T> input) {
+      input.apply(ParDo.of(new GridFsWriteFn<T>(this)));
+      return PDone.in(input.getPipeline());
+    }
+  }
+  private static class GridFsWriteFn<T> extends DoFn<T, Void> {
+
+    private final Write<T> spec;
+
+    private Mongo mongo;
+    private GridFS gridfs;
+
+    private GridFSInputFile gridFsFile;
+    private OutputStream outputStream;
+
+    public GridFsWriteFn(Write<T> spec) {
+      this.spec = spec;
+    }
+
+    private Mongo setupMongo() {
+      return spec.uri() == null ? new Mongo() : new Mongo(new MongoURI(spec.uri()));
+    }
+
+    private GridFS setupGridFS(Mongo mongo) {
```

Comment:
This method could be in the `Connection` POJO used by both `Read` and `Write` according to the previous comment.